### ISSUE TITLE
Markdown files are not rendered correctly

### ISSUE THREAD

**Issue by PavelTurk** [apache/netbeans#6652]
### Apache NetBeans version

Apache NetBeans 19

### What happened

This is my markdown file:

```
# Header 1
sdsdsds
## Header 2
sdsdsds
### Header 3
```

According this this [guide](https://www.markdownguide.org/basic-syntax/) it must be rendered this way:

![Screenshot from 2023-11-02 13-01-47](https://github.com/apache/netbeans/assets/10261361/22ae10b2-fe63-4459-89a0-22897d0080ea)

However, this is what I have in NB 19:

![Screenshot from 2023-11-02 13-03-16](https://github.com/apache/netbeans/assets/10261361/76835c64-1eaa-40a0-ab75-7963968c7810)

Please, note, that header 2 font size is even bigger that header 1 font size. 

Besides it would be nice if it were possible to edit CSS rules for rendered markdown files.

### Did this work correctly in an earlier version?

No / Don't know

### Operating System

Ubuntu 20.04.3 LTS

### JDK

OpenJDK 64-Bit Server VM (build 21.0.1+12-29, mixed mode, sharing)

### Apache NetBeans packaging

Apache NetBeans binary zip

### Are you willing to submit a pull request?

No

**Comment by lkishalmi** [comment]
> Maybe it is better to use a library like https://github.com/commonmark/commonmark-java to render markdown to html and make the preview html viewer? Besides it is possible to add rules.css file, that can be edited by NB user.

Well, we already doing that, using flexmark, as commonmark is no longer in development. The original effort is was on better than nothing base, refinements are in progress, PR-s are welcome.

**Comment by matthiasblaesing** [comment]
Something is fishy here. I looked into #6662 and did not see a real difference. What I see is this (Screenshot was taken with Amazon Corretto 17.0.9 on Ubuntu Linux):

![grafik](https://github.com/apache/netbeans/assets/2179736/0139c57f-ca79-4222-86c2-96f024d0de7b)

Yes, this could look better, but it is also not obviously wrong. Given this is the default stylesheet (at least from my reading):

https://github.com/openjdk/jdk/blob/master/src/java.desktop/share/classes/javax/swing/text/html/default.css

The result looks fitting. The right question IMHO is: what is wrong, that the `h1` is rendered so wrong?

**Comment by lkishalmi** [comment]
AFAIK NetBeans installs it's own `default.css` which is based on the LaF. Anyway, with some help, #6662 should take care of this problem for NB21.

**Comment by matthiasblaesing** [comment]
Before we make a buggy situation worse by adding complexity, we should fix the problem. What I see is, that there are multiple places in NetBeans that modify the shared stylesheet and that is the problem. There can be an argument for the LAF to modify the global stylesheet, but all other places should use the solution you suggest in the `MarkdownEditorKit`, i.e. override getStylesheet and do their modifications there and then.
So I think #6662 goes into the wrong direction by adding complexity without clean benefits.

**Comment by Chris2011** [comment]
So as I understand it correctly, I should rather not work on the PR?

**Comment by matthiasblaesing** [comment]
From my POV it depends on what problem should be fixed. The problem h1-header is displayed smaller than h2/h2-header, then I think we need to fix it in a different way (that problem would only be masked by a stylesheet specifically for h1). If the problem is "the rendering of the markdown preview looks ugly", then @lkishalmi approach to provide a special stylesheet or this editorkit is the right approach.

From a bikesheedding perspective I would see three stylesheets:
- one base stylesheet for dark themes
- one base stylesheet for light themes
- one user configurable style sheet overring the corresponding base stylesheet

The base stylesheets would then need to cover the elements flexmark could render. In the example from my screenshot I would expect a white background and the headers to be black. The gray looks strange for text display and blue for headers is also uncommon from my POV.

**Comment by lkishalmi** [comment]
I also think that it worth to check/fix other stylesheet change usages which are might mess up the system default StyleSheet, then re-evaluate this one.

Thanks for the hint @matthiasblaesing !

**Comment by Chris2011** [comment]
Ok, so maybe @lkishalmi and me come together and talk about the possibilities :). Thx @matthiasblaesing

**Comment by Chris2011** [comment]
I didn't know, that this view renders MD, I just noticed HTML. But I guess this is out of scope of this ticket here. Here we are just talking about opening a MD file in the editor and click the preview next to it. Not this view here. But maybe I'm wrong.

**Comment by Chris2011** [comment]
Got it, thx for the clarification.

**Comment by haidubogdan** [comment]
@lkishalmi @Chris2011 

I'm re-activating this topic :) as I'm thinking of implementing a quick patch.
Do you think that today we need to have configurable markdown css ?
If not I can prepare some small improvements PR with simple hardcoded css.
I have this proof of concept branch : https://github.com/apache/netbeans/compare/master...haidubogdan:netbeans:t_markdown_improvements

I see that there are a lot of limitation with css and swing html renderer, so maybe a point  would be to have a more reliable java html renderer.

before:

<img width="453" height="340" alt="Image" src="https://github.com/user-attachments/assets/175bae1b-7137-43f9-a981-c540544fd863" />

after : 

<img width="655" height="508" alt="Image" src="https://github.com/user-attachments/assets/43b14dc1-c35f-4f2e-b71d-d4c27082743a" />

@mbien I'm not very familiar with nb binaries there is a specific jar which deals with strikehrough elments rendering : flexmark-ext-gfm-strikethrough-0.64.8.jar . Maybe it can be added.

**Comment by mbien** [comment]
> @mbien I'm not very familiar with nb binaries there is a specific jar which deals with strikehrough elments rendering : flexmark-ext-gfm-strikethrough-0.64.8.jar . Maybe it can be added.

if its useful it certainly could be added. The dependency handling of NB modules can be indeed tricky since it entails some manual editing. Easiest way to approach it is to look at past updates, e.g in case of flexmark:

https://github.com/apache/netbeans/pull/6918/changes/ae5a15bd04f684c46e64da56a170e40b21408782 this would be a lib update, adding a jar would be a slice of the changeset.

we have to keep an eye on flexmark and if it is still receiving updates. JDK's javadoc markdown impl uses https://github.com/commonmark/commonmark-java which could be a candidate if needed.

> Do you think that today we need to have configurable markdown css ?

This could be a followup if needed but I think there is enough room to improve default rendering before making it customizable. Screenshots look good IMO!

**Comment by Chris2011** [comment]
@haidubogdan thx for your work, I'm wirh @mbien. Let's do it step by step, first your patch with hardcoded css and later one, if wanted, customizable css.

**Comment by Chris2011** [comment]
@haidubogdan there is also a PR where you can have a look or just create a new one :) https://github.com/apache/netbeans/pull/6662

**Comment by haidubogdan** [comment]
> [@haidubogdan](https://github.com/haidubogdan) there is also a PR where you can have a look or just create a new one :) [#6662](https://github.com/apache/netbeans/pull/6662)

I know, but from the comments I saw that there was a bottleneck regarding css customization.
Still, I think that even with css customization there is no direct translation for css rules using Swing Html Renderer, so it might not be that meaningful.
 Ex: `<hr>` elements don't take any css rules, `<code>` elements dont' apply padding ... a lot of limitations.

I will create a separate PR and I'l wait to see if there are some recommendation to include some of the 6662 proposals.

**Comment by Chris2011** [comment]
> > [@haidubogdan](https://github.com/haidubogdan) there is also a PR where you can have a look or just create a new one :) [#6662](https://github.com/apache/netbeans/pull/6662)
> 
> I know, but from the comments I saw that there was a bottleneck regarding css customization. Still, I think that even with css customization there is no direct translation for css rules using Swing Html Renderer, so it might not be that meaningful. Ex: `<hr>` elements don't take any css rules, `<code>` elements dont' apply padding ... a lot of limitations.
> 
> I will create a separate PR and I'l wait to see if there are some recommendation to include some of the 6662 proposals.

True, I also guess that a lot of new CSS features can't handle by the Swing HTML renderer. But just guessing.
